kafka-client/src/main/java/dev/responsive/kafka/api/async/internals/AsyncThreadPool.java

```diff
@@ -135,21 +142,36 @@ public <KOut, VOut> void scheduleForProcessing(
     final var inFlightKey = InFlightWorkKey.of(processorName, partition);
     final var inFlightForTask
         = inFlight.computeIfAbsent(inFlightKey, k -> new ConcurrentHashMap<>());
-    for (final var e : events) {
+
+    for (final AsyncEvent event : events) {
       final var future = CompletableFuture.runAsync(
-          new AsyncEventTask<>(e, taskContext, asyncProcessorContext, finalizingQueue),
+          new AsyncEventTask<>(event, taskContext, asyncProcessorContext),
           executor
       );
       final var inFlightEvent = new InFlightEvent(future);
-      inFlightForTask.put(e, inFlightEvent);
-      future.handle((r, t) -> {
-        if (t != null) {
-          inFlightEvent.setError(t);
-        } else {
-          inFlightForTask.remove(e);
-        }
-        return null;
-      });
+      inFlightForTask.put(event, inFlightEvent);
+
+      future
```

Comment:
This refactoring of the future/adding stages is one of the main changes in this PR so definitely take a close look.

Reply:
makes sense to me!